# Print phy vs write fail; retry the 15:50:30 DDR3 TPR at 1.3 V then 1.1 V.
#
--- a/arch/arm/mach-sunxi/dram_sun50i_h616.c
+++ b/arch/arm/mach-sunxi/dram_sun50i_h616.c
@@ -1362,6 +1362,8 @@
 	return mctl_ctrl_init(para, config);
 }
 
+static bool walnutpi_any_phy;
+
 static bool walnutpi_sdram_works(void)
 {
 	writel(0x5a5aa5a5, CFG_SYS_SDRAM_BASE);
@@ -1377,7 +1379,10 @@
 static bool walnutpi_rank_ok(const struct dram_para *para,
 			     struct dram_config *config)
 {
-	return mctl_core_init(para, config) && walnutpi_sdram_works();
+	if (!mctl_core_init(para, config))
+		return false;
+	walnutpi_any_phy = true;
+	return walnutpi_sdram_works();
 }
 
 static bool mctl_auto_detect_rank_width(const struct dram_para *para,
@@ -1494,14 +1499,26 @@
 	para->tpr12 = tpr12;
 }
 
+static void walnutpi_load_ddr3_vendor(struct dram_para *para)
+{
+	para->type = SUNXI_DRAM_TYPE_DDR3;
+	para->clk = 792;
+	para->dx_odt = 0x07070707;
+	para->tpr6 = 0x4400c080;
+	para->tpr10 = 0x40ef6663;
+	para->tpr11 = 0x1a1a1a16;
+	para->tpr12 = 0x14121210;
+}
+
 static bool walnutpi_try(const char *tag, struct dram_para *para,
 			 struct dram_config *config)
 {
+	walnutpi_any_phy = false;
 	if (mctl_auto_detect_rank_width(para, config)) {
 		printf(" %s+", tag);
 		return true;
 	}
-	printf(" %s-", tag);
+	printf(" %s-%c", tag, walnutpi_any_phy ? 'w' : 'p');
 	return false;
 }
 
@@ -1541,8 +1558,14 @@
 	walnutpi_load_ddr3(&para, 792);
 	if (walnutpi_try("d3C", &para, &config))
 		goto detected;
+	walnutpi_load_ddr3_vendor(&para);
+	if (walnutpi_try("d3D", &para, &config))
+		goto detected;
 
 	walnutpi_dram_volt(1100);
+	walnutpi_load_ddr3_vendor(&para);
+	if (walnutpi_try("d3E", &para, &config))
+		goto detected;
 	walnutpi_load_lpddr4(&para, 0x4400c080, 0x40ef6663, 0x24242323, 0x0e0e0e0e);
 	if (walnutpi_try("lp4A", &para, &config))
 		goto detected;
